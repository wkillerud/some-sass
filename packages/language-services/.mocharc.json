{
	"ui": "tdd",
	"color": true,
	"spec": "./dist/**/*.test.js",
	"recursive": true
}
