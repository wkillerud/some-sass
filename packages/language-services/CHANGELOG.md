## 1.3.4 (2024-08-24)


### ğŸ©¹ Fixes

- race condition in workspace scanner

- exclude sass globals if suggestFromUseOnly


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.3.1

## 1.3.3 (2024-08-20)


### ğŸ©¹ Fixes

- a race condition that caused partials not to be parsed sometimes

## 1.3.2 (2024-08-20)


### ğŸ©¹ Fixes

- apply completion settings

## 1.3.1 (2024-08-11)


### ğŸ©¹ Fixes

- adjust when suggesting functions, variables in control flow

- add prefix support to forwarded sass built-ins

## 1.3.0 (2024-08-07)


### ğŸš€ Features

- feature update for language server


### ğŸ©¹ Fixes

- suggest from forwarded sass built-ins


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.3.0

## 1.2.0 (2024-08-04)


### ğŸš€ Features

- feature update for language server


### ğŸ©¹ Fixes

- update dependency vscode-languageserver-textdocument to v1.0.12

- suggest variables and functions as mixin parameters

- functions in string interpolation get suggestions

- **deps:** update dependency scss-sassdoc-parser to v3.2.0


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.2.0

## 1.1.0 (2024-07-30)


### ğŸš€ Features

- add support for loadPaths


### ğŸ©¹ Fixes

- bugfix update for language server


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.1.0

## 1.0.8 (2024-07-29)


### ğŸ©¹ Fixes

- update language server to fix pkg resolution bug

- completions after warn, error and debug statements

- mark both variable and function contexts in interpolation

- completions bug where symbols in first document weren't included

- support suggestions in interpolation in class selector

- **deps:** update dependency colorjs.io to v0.5.2


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.0.6

## 1.0.7 (2024-07-28)


### ğŸ©¹ Fixes

- update language server to fix #182

- update vscode-css-languageservice


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.0.5

## 1.0.6 (2024-07-25)


### ğŸ©¹ Fixes

- update language server for completion regression fixes


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.0.4

## 1.0.5 (2024-06-16)


### ğŸ©¹ Fixes

- update language server for completions bugfix

- hide private symbols with - prefix

- only suggest from current namespace

- for wildcard use, omit . for functions in filtertext


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.0.3

## 1.0.4 (2024-06-15)


### ğŸ©¹ Fixes

- read setting for maximum color decorators ([25d276b5](https://github.com/wkillerud/some-sass/commit/25d276b5))

- off-by-one for the start of namespace traversal ([#171](https://github.com/wkillerud/some-sass/pull/171))


### ğŸ§± Updated Dependencies

- Updated @somesass/vscode-css-languageservice to 1.0.2

## 1.0.3 (2024-05-30)


### ğŸ©¹ Fixes

- look up the editor limit for color decorators ([#165](https://github.com/wkillerud/some-sass/pull/165))


### ğŸ§± Updated Dependencies

- Updated some-sass-language-server to 1.2.4

## 1.0.2 (2024-05-01)


### ğŸ©¹ Fixes

- update vscode-css-languageservice to 6.2.14