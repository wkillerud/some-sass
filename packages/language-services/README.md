# @somesass/language-services

Experimental wrapper around `@somesass/vscode-css-languageservice`.
