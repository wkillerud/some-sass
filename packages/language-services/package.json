{
	"name": "@somesass/language-services",
	"version": "2.0.1",
	"private": true,
	"description": "The features powering some-sass-language-server",
	"keywords": [
		"scss",
		"sass"
	],
	"engines": {
		"node": ">=20"
	},
	"homepage": "https://github.com/wkillerud/some-sass/blob/main/packages/language-services#readme",
	"repository": {
		"type": "git",
		"url": "git+ssh://git@github.com/wkillerud/some-sass.git"
	},
	"bugs": {
		"url": "https://github.com/wkillerud/some-sass/issues"
	},
	"files": [
		"dist/",
		"!dist/test/",
		"!dist/**/*.test.js"
	],
	"main": "dist/language-services.js",
	"types": "dist/language-services.d.ts",
	"exports": {
		".": {
			"types": "./dist/language-services.d.ts",
			"default": "./dist/language-services.js"
		},
		"./*": {
			"types": "./dist/*.d.ts",
			"default": "./dist/*.js"
		},
		"./feature/*": {
			"types": "./dist/feature/*.d.ts",
			"default": "./dist/feature/*.js"
		}
	},
	"author": "William Killerud <william@killerud.com> (https://www.williamkillerud.com/)",
	"license": "MIT",
	"scripts": {
		"build": "tsc",
		"clean": "shx rm -rf dist node_modules",
		"test": "vitest",
		"coverage": "vitest run --coverage"
	},
	"dependencies": {
		"@somesass/vscode-css-languageservice": "2.0.0",
		"@types/lodash.merge": "4.6.9",
		"colorjs.io": "0.5.2",
		"lodash.merge": "4.6.2",
		"sassdoc-parser": "3.4.1"
	},
	"devDependencies": {
		"@vitest/coverage-v8": "2.0.5",
		"shx": "0.3.4",
		"typescript": "5.5.4"
	}
}
