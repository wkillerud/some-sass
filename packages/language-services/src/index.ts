export * from "./language-services";
export * from "./language-services-types";
