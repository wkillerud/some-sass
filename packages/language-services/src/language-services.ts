import type {
	LanguageService,
	LanguageServiceOptions,
	TextDocument,
} from "@somesass/language-server-types";
import { SassLinkFinder } from "./feature/find-links";
import { SassSymbolFinder } from "./feature/find-symbols";
import { getLanguageModelCache } from "./language-model-cache";

export { getLanguageModelCache };

export function getLanguageService(
	options: LanguageServiceOptions,
): LanguageService {
	const linkFinder = new SassLinkFinder(options);
	const symbolFinder = new SassSymbolFinder(options);

	return {
		configure: (settings) => {
			linkFinder.configure(settings);
		},
		parseStylesheet: (document: TextDocument) =>
			options.languageModelCache.get(document),
		findDocumentLinks: linkFinder.findDocumentLinks.bind(linkFinder),
		findDocumentSymbols: symbolFinder.findDocumentSymbols.bind(symbolFinder),
	};
}
