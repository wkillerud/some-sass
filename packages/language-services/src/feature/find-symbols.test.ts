import * as assert from "node:assert";
import {
	LanguageService,
	SassDocumentSymbol,
	SymbolKind,
	SyntaxNodeType,
	TextDocument,
} from "@somesass/language-server-types";
import { getLanguageService } from "../language-services";
import { newRange } from "../test/test-resources";

const getLS = () => getLanguageService();

export function assertDocumentSymbols(
	ls: LanguageService,
	input: string,
	expected: SassDocumentSymbol[],
	lang: string = "css",
) {
	const document = TextDocument.create(
		`test://test/test.${lang}`,
		lang,
		0,
		input,
	);

	const stylesheet = ls.parseStylesheet(document);

	const symbols = ls.findDocumentSymbols(document, stylesheet);
	assert.deepEqual(symbols, expected);
}

test("basic document symbols", () => {
	const ls = getLS();
	assertDocumentSymbols(ls, ".foo {}", [
		{
			name: ".foo",
			type: SyntaxNodeType.ClassName,
			kind: SymbolKind.Class,
			range: newRange(0, 7),
			selectionRange: newRange(0, 4),
		},
	]);
	assertDocumentSymbols(ls, ".foo:not(.selected) {}", [
		{
			name: ".foo:not(.selected)",
			type: SyntaxNodeType.ClassName,
			kind: SymbolKind.Class,
			range: newRange(0, 22),
			selectionRange: newRange(0, 19),
		},
	]);

	// multiple selectors, each range starts with the selector offset
	assertDocumentSymbols(ls, ".voo.doo, .bar {}", [
		{
			name: ".voo.doo",
			kind: SymbolKind.Class,
			type: SyntaxNodeType.ClassName,
			range: newRange(0, 17),
			selectionRange: newRange(0, 8),
		},
		{
			name: ".bar",
			kind: SymbolKind.Class,
			type: SyntaxNodeType.ClassName,
			range: newRange(10, 17),
			selectionRange: newRange(10, 14),
		},
	]);

	// Media Query
	assertDocumentSymbols(ls, "@media screen, print {}", [
		{
			name: "@media screen, print",
			kind: SymbolKind.Module,
			type: SyntaxNodeType.MediaStatement,
			range: newRange(0, 23),
			selectionRange: newRange(7, 20),
		},
	]);
	assertDocumentSymbols(ls, "@media screen, print {}", [
		{
			name: "@media screen, print",
			type: SyntaxNodeType.MediaStatement,
			kind: SymbolKind.Module,
			range: newRange(0, 23),
			selectionRange: newRange(7, 20),
		},
	]);
});

test("mixins", () => {
	const ls = getLS();
	assertDocumentSymbols(ls, "@mixin foo { }", [
		{
			name: "foo",
			type: SyntaxNodeType.MixinStatement,
			kind: SymbolKind.Method,
			range: newRange(0, 14),
			selectionRange: newRange(7, 10),
		},
	]);
	assertDocumentSymbols(ls, "@mixin {}", [
		{
			name: "<undefined>",
			type: SyntaxNodeType.MixinStatement,
			kind: SymbolKind.Method,
			range: newRange(0, 9),
			selectionRange: newRange(0, 0),
		},
	]);
});
