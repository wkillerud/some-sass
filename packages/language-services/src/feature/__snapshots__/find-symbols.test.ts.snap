// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`all (at least most of) the things - sass 1`] = `[]`;

exports[`all (at least most of) the things 1`] = `[]`;

exports[`basic document symbols - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".foo",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
]
`;

exports[`basic document symbols 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".foo",
    "range": {
      "end": {
        "character": 7,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
]
`;

exports[`complex selector - sass 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 5,
        "name": "not",
        "range": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 0,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 8,
            "line": 0,
          },
          "start": {
            "character": 5,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "PseudoClassName",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 5,
            "name": ".selected",
            "range": {
              "end": {
                "character": 19,
                "line": 0,
              },
              "start": {
                "character": 8,
                "line": 0,
              },
            },
            "selectionRange": {
              "end": {
                "character": 18,
                "line": 0,
              },
              "start": {
                "character": 9,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "ClassSelector",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "(.selected)",
        "range": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 0,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 8,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 5,
    "name": ".foo:not(.selected)",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 19,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "PseudoClassSelector",
  },
]
`;

exports[`complex selector 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 5,
        "name": "not",
        "range": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 0,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 8,
            "line": 0,
          },
          "start": {
            "character": 5,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "PseudoClassName",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 5,
            "name": ".selected",
            "range": {
              "end": {
                "character": 19,
                "line": 0,
              },
              "start": {
                "character": 8,
                "line": 0,
              },
            },
            "selectionRange": {
              "end": {
                "character": 18,
                "line": 0,
              },
              "start": {
                "character": 9,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "ClassSelector",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "(.selected)",
        "range": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 0,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 8,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 5,
    "name": ".foo:not(.selected)",
    "range": {
      "end": {
        "character": 22,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 19,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "PseudoClassSelector",
  },
]
`;

exports[`function declaration - sass 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 6,
        "name": "fibonacci",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "Callee",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 13,
            "name": "$n",
            "range": {
              "end": {
                "character": 23,
                "line": 0,
              },
              "start": {
                "character": 19,
                "line": 0,
              },
            },
            "sassdoc": undefined,
            "selectionRange": {
              "end": {
                "character": 22,
                "line": 0,
              },
              "start": {
                "character": 20,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "SassVariableName",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "($n)",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 19,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 12,
    "name": "fibonacci($n)",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 6,
        "name": "fibonacci",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "Callee",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 13,
            "name": "$n",
            "range": {
              "end": {
                "character": 23,
                "line": 0,
              },
              "start": {
                "character": 19,
                "line": 0,
              },
            },
            "sassdoc": undefined,
            "selectionRange": {
              "end": {
                "character": 22,
                "line": 0,
              },
              "start": {
                "character": 20,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "SassVariableName",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "($n)",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 19,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 6,
    "name": "fibonacci($n)",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "CallExpression",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 13,
    "name": "$n",
    "range": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 19,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 22,
        "line": 0,
      },
      "start": {
        "character": 20,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "SassVariableName",
  },
]
`;

exports[`function declaration 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 6,
        "name": "fibonacci",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "Callee",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 13,
            "name": "$n",
            "range": {
              "end": {
                "character": 23,
                "line": 0,
              },
              "start": {
                "character": 19,
                "line": 0,
              },
            },
            "sassdoc": undefined,
            "selectionRange": {
              "end": {
                "character": 22,
                "line": 0,
              },
              "start": {
                "character": 20,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "SassVariableName",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "($n)",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 19,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 12,
    "name": "fibonacci($n)",
    "range": {
      "end": {
        "character": 27,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
  {
    "children": [
      {
        "children": undefined,
        "detail": undefined,
        "kind": 6,
        "name": "fibonacci",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 19,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "Callee",
      },
      {
        "children": [
          {
            "children": undefined,
            "detail": undefined,
            "kind": 13,
            "name": "$n",
            "range": {
              "end": {
                "character": 23,
                "line": 0,
              },
              "start": {
                "character": 19,
                "line": 0,
              },
            },
            "sassdoc": undefined,
            "selectionRange": {
              "end": {
                "character": 22,
                "line": 0,
              },
              "start": {
                "character": 20,
                "line": 0,
              },
            },
            "tags": undefined,
            "type": "SassVariableName",
          },
        ],
        "detail": undefined,
        "kind": 6,
        "name": "($n)",
        "range": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 10,
            "line": 0,
          },
        },
        "selectionRange": {
          "end": {
            "character": 23,
            "line": 0,
          },
          "start": {
            "character": 19,
            "line": 0,
          },
        },
        "tags": undefined,
        "type": "ArgList",
      },
    ],
    "detail": undefined,
    "kind": 6,
    "name": "fibonacci($n)",
    "range": {
      "end": {
        "character": 27,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "CallExpression",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 13,
    "name": "$n",
    "range": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 19,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 22,
        "line": 0,
      },
      "start": {
        "character": 20,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "SassVariableName",
  },
]
`;

exports[`media query - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 2,
    "name": "@media screen, print
",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MediaStatement",
  },
]
`;

exports[`media query 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 2,
    "name": "@media screen, print {}",
    "range": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MediaStatement",
  },
]
`;

exports[`mixin - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 6,
    "name": "foo",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 10,
        "line": 0,
      },
      "start": {
        "character": 7,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
]
`;

exports[`mixin 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 6,
    "name": "foo",
    "range": {
      "end": {
        "character": 14,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 10,
        "line": 0,
      },
      "start": {
        "character": 7,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
]
`;

exports[`mixin without a name - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 6,
    "name": "<undefined>",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
]
`;

exports[`mixin without a name 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 6,
    "name": "<undefined>",
    "range": {
      "end": {
        "character": 9,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "MixinStatement",
  },
]
`;

exports[`multiple selectors for same block, each range starts with the selector offset - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".voo.doo",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 8,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".voo",
    "range": {
      "end": {
        "character": 8,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".bar",
    "range": {
      "end": {
        "character": 0,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 14,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
]
`;

exports[`multiple selectors for same block, each range starts with the selector offset 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".voo.doo",
    "range": {
      "end": {
        "character": 17,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 8,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".voo",
    "range": {
      "end": {
        "character": 8,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".bar",
    "range": {
      "end": {
        "character": 17,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 14,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
]
`;

exports[`placeholder selector - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": "%placeholder",
    "range": {
      "end": {
        "character": 0,
        "line": 3,
      },
      "start": {
        "character": 0,
        "line": 1,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 12,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 1,
      },
    },
    "tags": undefined,
    "type": "PlaceholderSelector",
  },
]
`;

exports[`placeholder selector 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": "%placeholder",
    "range": {
      "end": {
        "character": 29,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "PlaceholderSelector",
  },
]
`;

exports[`placeholder selector usage - sass 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".theme",
    "range": {
      "end": {
        "character": 0,
        "line": 3,
      },
      "start": {
        "character": 0,
        "line": 1,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 6,
        "line": 1,
      },
      "start": {
        "character": 0,
        "line": 1,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": "%theme",
    "range": {
      "end": {
        "character": 0,
        "line": 3,
      },
      "start": {
        "character": 1,
        "line": 2,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 15,
        "line": 2,
      },
      "start": {
        "character": 9,
        "line": 2,
      },
    },
    "tags": undefined,
    "type": "PlaceholderSelector",
  },
]
`;

exports[`placeholder selector usage 1`] = `
[
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": ".theme",
    "range": {
      "end": {
        "character": 26,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 6,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "ClassSelector",
  },
  {
    "children": undefined,
    "detail": undefined,
    "kind": 5,
    "name": "%theme",
    "range": {
      "end": {
        "character": 24,
        "line": 0,
      },
      "start": {
        "character": 9,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 23,
        "line": 0,
      },
      "start": {
        "character": 17,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "PlaceholderSelector",
  },
]
`;

exports[`variable - sass 1`] = `
[
  {
    "children": undefined,
    "detail": "lime",
    "kind": 13,
    "name": "$color",
    "range": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 6,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "SassVariableName",
  },
]
`;

exports[`variable 1`] = `
[
  {
    "children": undefined,
    "detail": "lime",
    "kind": 13,
    "name": "$color",
    "range": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "sassdoc": undefined,
    "selectionRange": {
      "end": {
        "character": 6,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "tags": undefined,
    "type": "SassVariableName",
  },
]
`;
